File: src/task.go
package main

import (
	"log"
	"os"
	"sync"
)

type TaskStatus int

const (
	QUEUED = iota
	PROCESSING
	DONE
	FAILED
)

type Task struct {
	Path       string
	OutputPath string
	Probe      *ProbeResult
	Progress   *TranscodingProgress
	WebhookId  *string
	Status     TaskStatus
}

func TaskNew(path string, outputPath string) Task {
	return Task{
		Path:       path,
		OutputPath: outputPath,
		Status:     QUEUED,
	}
}

func (task *Task) run(mu *sync.Mutex) error {
	discord := NewDiscordClient(os.Getenv("DISCORD_WEBHOOK"))
	discord.Trigger(task)
	mu.Lock()
	defer mu.Unlock()
	task.Status = PROCESSING
	err := ffprobe(task)
	if err != nil {
		log.Println("ERR: ffprobe failed for file: " + task.Path)
		task.Status = FAILED
		discord.Trigger(task)
		return err
	}
	discord.Trigger(task)
	err = ffmpeg(task, func() {
		discord.Trigger(task)
	})
	if err != nil {
		log.Println("ERR: ffmpeg failed for file: " + task.Path)
		task.Status = FAILED
		discord.Trigger(task)
		return err
	}
	_ = os.Remove(task.Path)
	task.Status = DONE
	discord.Trigger(task)
	return nil
}

File: api/src/services/task.rs
use lib::{
    discord::{DiscordEventHandler, DiscordWebhook},
    ffmpeg::FFMpeg,
    ffprobe::ffprobe,
};
use log::{error, info};
use std::{path::Path, sync::Arc};
use tokio::{sync::Mutex, task::JoinSet};

use crate::state::AppArgs;

pub struct TaskService {
    mutex: Arc<Mutex<()>>,
    args: Arc<AppArgs>,
}

impl TaskService {
    pub fn new(args: Arc<AppArgs>) -> Self {
        Self {
            mutex: Arc::new(Mutex::new(())),
            args,
        }
    }

    pub async fn run_task(&self, input_path: &Path, output_path: &Path) {
        info!("Transcoding: {:?} to {:?}", input_path, output_path);

        let probe = ffprobe(&input_path).await.expect("ffprobe failed");
        let mut ffmpeg = FFMpeg::new(&self.args.config.ffmpeg);
        let mut join_set = JoinSet::new();

        if let Some(webhook_url) = &self.args.config.discord.webhook_url {
            let webhook = DiscordWebhook::new(&webhook_url);
            let mut discord_handler = DiscordEventHandler::new(webhook);
            let rx = ffmpeg.subscribe();
            join_set.spawn(async move {
                discord_handler.listen(rx).await;
            });
        }

        let mut _guard = self.mutex.lock().await;

        if let Err(e) = ffmpeg.transcode(&probe, output_path).await {
            error!("An error happened while transcoding {}", e);
        }

        join_set.join_all().await;
    }
}
